### PR TITLE
mds : optimization functions,get_dirfrags_under, to speed up processing directories with tens of millions of files

### PR DESCRIPTION
**Pull request by simon-rock**
it now takes 60ms to process a frag of dir(have tens of millions of files)

Signed-off-by: simon gao <simon29rock@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/mds/CInode.cc [hunk 1]
```diff
@@ -703,41 +703,35 @@ std::pair<bool, std::vector<CDir*>> CInode::get_dirfrags_under(frag_t fg)
...
-      } else {
-        all = false;
-      }
-    }
-  }
-
-  if (all)
+  all = false;
+  frag_vec_t children;
+  if (dirfrags.count(fg)){
```
Comment (ukernel) [inline src/mds/CInode.cc]: if (auto it = dirfrags.find(fg); it != dirfrags.end()) {
   dirs.push_back(it->second());
   return result;
}

File: src/mds/CInode.cc [hunk 2]
```diff
@@ -703,41 +703,35 @@ std::pair<bool, std::vector<CDir*>> CInode::get_dirfrags_under(frag_t fg)
...
-    tmpdft.force_to_leaf(g_ceph_context, p.first);
-    if (fg.contains(p.first) && !dirfragtree.is_leaf(p.first))
+  }
+  for(auto &p : dirfrags){
+    if (fg.bits() < p.first.bits() && fg.contains(p.first)){
+      children.push_back(p.first);
       dirs.push_back(p.second);
+    }else if (fg.bits() > p.first.bits() && p.first.contains(fg)){
+      all = true;
+      break;
```
Comment (ukernel) [inline src/mds/CInode.cc]: return result;

File: src/mds/CInode.cc [hunk 3]
```diff
@@ -703,41 +703,35 @@ std::pair<bool, std::vector<CDir*>> CInode::get_dirfrags_under(frag_t fg)
...
-  tmpdft.force_to_leaf(g_ceph_context, fg);
-  for (auto &p : dirfrags) {
-    tmpdft.force_to_leaf(g_ceph_context, p.first);
-    if (fg.contains(p.first) && !dirfragtree.is_leaf(p.first))
+  }
+  for(auto &p : dirfrags){
+    if (fg.bits() < p.first.bits() && fg.contains(p.first)){
+      children.push_back(p.first);
       dirs.push_back(p.second);
+    }else if (fg.bits() > p.first.bits() && p.first.contains(fg)){
```
Comment (ukernel) [inline src/mds/CInode.cc]: } else if

File: src/mds/CInode.cc [hunk 4]
```diff
@@ -703,41 +703,35 @@ std::pair<bool, std::vector<CDir*>> CInode::get_dirfrags_under(frag_t fg)
...
-        all = false;
-      }
-    }
-  }
-
-  if (all)
+  all = false;
+  frag_vec_t children;
+  if (dirfrags.count(fg)){
+    all = true;
```
Comment (ukernel) [inline src/mds/CInode.cc]:
```
std::pair<bool, std::vector<CDir*>> CInode::get_dirfrags_under(frag_t fg)
{
  std::pair<bool, std::vector<CDir*>> result;
  auto& all = result.first;
  auto& dirs = result.second;

  if (auto it = dirfrags.find(fg); it != dirfrags.end()){
    all = true;
    dirs.push_back(it->second);
    return result;
  }

  for(auto &p : dirfrags){
    if (fg.bits() >= p.first.bits() && p.first.contains(fg)) {
      all = true;
      return result;
    }
    if (fg.bits() < p.first.bits() && fg.contains(p.first))
      dirs.push_back(p.second);
  }

  if (!dirs.empty()) {
    frag_vec_t leaves;
    dirfragtree.get_leaves_under(fg, leaves);
    if (leaves.size() == dirs.size())
      all = true;
  }

  return result;
}
```

**Comment by ukernel** [comment]
I think something like below is cleaner

```
std::vector<CDir*> CInode::get_dirfrags_under(frag_t fg, bool* all )
{
  std::vector<CDir*> dirs;

  if (auto it = dirfrags.find(fg); it != dirfrags.end()){
    if(all)
      *all = true;
    dirs.push_back(it->second);
    return dirs;
  }

  int64_t total = 0;
  for(auto &p : dirfrags) {
   if (fg.bits() >= p.first.bits() && p.first.contains(fg)) {
     if (*all)
        all = true;
     return dirs;
   }
   if (fg.bits() < p.first.bits() && fg.contains(p.first)){
      dirs.push_back(p.second);
      total += 1 << (24 - p.first.bits());
    }
  }

  if (all)
    *all = (1<<(24-fg.bits())) == total) 
  return dirs;
}
```

File: src/mds/CInode.cc [hunk 5]
```diff
@@ -697,48 +697,35 @@ frag_t InodeStoreBase::pick_dirfrag(std::string_view dn)
...
-
-  if (all)
-    return result;
-
-  fragtree_t tmpdft;
-  tmpdft.force_to_leaf(g_ceph_context, fg);
-  for (auto &p : dirfrags) {
-    tmpdft.force_to_leaf(g_ceph_context, p.first);
-    if (fg.contains(p.first) && !dirfragtree.is_leaf(p.first))
+  int64_t total = 0;
```
Comment (ukernel) [inline src/mds/CInode.cc]: no need to use int64 here

File: src/mds/CInode.cc [hunk 6]
```diff
@@ -697,48 +697,36 @@ frag_t InodeStoreBase::pick_dirfrag(std::string_view dn)
   return dirfragtree[h];
 }
 
-std::pair<bool, std::vector<CDir*>> CInode::get_dirfrags_under(frag_t fg)
+std::vector<CDir*> CInode::get_dirfrags_under(frag_t fg, bool* all )
 {
-  std::pair<bool, std::vector<CDir*>> result;
-  auto& all = result.first;
-  auto& dirs = result.second;
+  std::vector<CDir*> dirs;
```
Comment (batrick) [inline src/mds/CInode.cc]: I don't really see a reason to change the return type. Can you please revert this.

Comment (batrick) [reply src/mds/CInode.cc]: Also: `all` should always be true unless we're returning from the end of this method. Might as well set it `true` here.

File: src/mds/CInode.cc [hunk 7]
```diff
@@ -697,48 +697,36 @@ frag_t InodeStoreBase::pick_dirfrag(std::string_view dn)
...
-      if (!dirfrags.count(leaf)) {
-        all = false;
-        break;
+  
+  int total = 0;
+  for(auto &p : dirfrags) {
+    if (fg.bits() >= p.first.bits()) {
+      if (p.first.contains(fg)) {
+	if (all)
+	  *all = true;
```
Comment (batrick) [inline src/mds/CInode.cc]: Forgot to include `p` in results?

File: src/mds/CInode.cc [hunk 8]
```diff
@@ -697,48 +697,36 @@ frag_t InodeStoreBase::pick_dirfrag(std::string_view dn)
...
-  {
-    frag_vec_t leaves;
-    tmpdft.get_leaves_under(fg, leaves);
-    for (const auto& leaf : leaves) {
-      if (!dirfrags.count(leaf)) {
-        all = false;
-        break;
+  
+  int total = 0;
+  for(auto &p : dirfrags) {
```
Comment (batrick) [inline src/mds/CInode.cc]: `auto& [fg, dir] ...`

to simplify

File: src/mds/CInode.cc [hunk 9]
```diff
@@ -697,48 +697,36 @@ frag_t InodeStoreBase::pick_dirfrag(std::string_view dn)
...
+    if (fg.bits() >= p.first.bits()) {
```
Comment (batrick) [inline src/mds/CInode.cc]: Please comment these cases.

File: src/mds/CInode.cc [hunk 10]
```diff
@@ -697,48 +697,36 @@ frag_t InodeStoreBase::pick_dirfrag(std::string_view dn)
...
+      if (p.first.contains(fg)) {
+	if (all)
+	  *all = true;
+	return dirs;
+      }
+    } else {
+      if (fg.contains(p.first)) {
+	dirs.push_back(p.second);
+	total += 1 << (24 - p.first.bits());
```
Comment (batrick) [inline src/mds/CInode.cc]: @ukernel these are magic numbers. Either add a comment or suggest a helper method in `frag_t`?

File: src/mds/CInode.cc (hunk 7 above)
Comment (simon-rock) [reply src/mds/CInode.cc]: In this case, the original output of this function does not contain this partition.

Comment (simon-rock) [reply src/mds/CInode.cc]: > Forgot to include `p` in results?

In this case, the original output of this function does not contain this partition, I think

File: src/mds/CInode.cc (hunk 6 above)
Comment (simon-rock) [reply src/mds/CInode.cc]: en. 

> I don't really see a reason to change the return type. Can you please revert this.
At the beginning of the period, I expect to set the return value of all as optional. This return value is only used in directory consolidation calculation. In the case, there is a more optimized method to calculate it.  I will revert this.

Comment (simon-rock) [reply src/mds/CInode.cc]: > Also: `all` should always be true unless we're returning from the end of this method. Might as well set it `true` here.

The return value of "all" means whether the partition in memory completely contains the input partition. The original method is to find whether the partition is not completely contained in memory, so firstly set it to true. Now the improvement The method is to calculate whether the current memory contains this partition completely, so the default is incomplete firstly

File: src/mds/CInode.cc (hunk 10 above)
Comment (simon-rock) [reply src/mds/CInode.cc]: I have updated it.

File: src/mds/CInode.cc (hunk 9 above)
Comment (simon-rock) [reply src/mds/CInode.cc]: updated

File: src/mds/CInode.cc (hunk 8 above)
Comment (simon-rock) [reply src/mds/CInode.cc]: updated

**Pull request merged**